core/src/plugins/filed/grpc/CMakeLists.txt

```diff
@@ -0,0 +1,138 @@
+#   License as published by the Free Software Foundation and included
+#   in the file LICENSE.
+#
+#   This program is distributed in the hope that it will be useful, but
+#   WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+#   Affero General Public License for more details.
+#
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program; if not, write to the Free Software
+#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+#   02110-1301, USA.
+
+include(BareosDisableWarnings)
+include(BareosConfigureFile)
+
+if(HAVE_WIN32)
+  message(STATUS "Disabling grpc plugin as its not supported on windows")
+elseif(HAVE_GRPC)
+  message(STATUS "Using protobuf ${Protobuf_VERSION}")
+  message(STATUS "Using gRPC ${gRPC_VERSION}")
+
+  set(_PROTOBUF_LIBPROTOBUF protobuf::libprotobuf)
+  if(TARGET gRPC::grpc++_reflection)
+    set(_REFLECTION gRPC::grpc++_reflection)
+  else()
+    set(_REFLECTION grpc++_reflection)
+  endif()
+  if(CMAKE_CROSSCOMPILING)
+    find_program(_PROTOBUF_PROTOC protoc)
+  else()
+    set(_PROTOBUF_PROTOC $<TARGET_FILE:protobuf::protoc>)
+  endif()
+  if(TARGET gRPC::grpc++)
+    set(_GRPC_GRPCPP gRPC::grpc++)
+  else()
+    set(_GRPC_GRPCPP grpc++)
+  endif()
+  if(TARGET gRPC::grpc_cpp_plugin)
+    set(_GRPC_CPP_PLUGIN_EXECUTABLE $<TARGET_FILE:gRPC::grpc_cpp_plugin>)
+  else()
+    find_program(_GRPC_CPP_PLUGIN_EXECUTABLE grpc_cpp_plugin)
+  endif()
+
+  # Proto file
+  macro(addgrpcprotofiles protosrcs protohdrs grpcsrcs grpchdrs)
+    foreach(file ${ARGN})
+      get_filename_component(absolute "${file}" ABSOLUTE)
+      get_filename_component(path "${absolute}" PATH)
+      get_filename_component(stem "${file}" NAME_WE)
+
+      set(proto_src "${CMAKE_CURRENT_BINARY_DIR}/${stem}.pb.cc")
+      set(proto_hdr "${CMAKE_CURRENT_BINARY_DIR}/${stem}.pb.h")
+      set(grpc_src "${CMAKE_CURRENT_BINARY_DIR}/${stem}.grpc.pb.cc")
+      set(grpc_hdr "${CMAKE_CURRENT_BINARY_DIR}/${stem}.grpc.pb.h")
+        if(Protobuf_VERSION VERSION_LESS 3.15.0)
+          set(_PROTOBUF_PROTOC_EXTRA_PARAMS "--experimental_allow_proto3_optional")
+        else()
+          set(_PROTOBUF_PROTOC_EXTRA_PARAMS "")
+        endif()
+      add_custom_command(
+        OUTPUT "${proto_src}" "${proto_hdr}" "${grpc_src}" "${grpc_hdr}"
+        COMMAND
+          ${_PROTOBUF_PROTOC} ARGS --grpc_out "${CMAKE_CURRENT_BINARY_DIR}"
+          --cpp_out "${CMAKE_CURRENT_BINARY_DIR}" -I "${path}"
+          --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+          ${_PROTOBUF_PROTOC_EXTRA_PARAMS}
+          "${absolute}"
+        DEPENDS "${absolute}"
+      )
+
+      list(APPEND protosrcs "${proto_src}")
+      list(APPEND protohdrs "${proto_hdr}")
+      list(APPEND grpcsrcs "${grpc_src}")
+      list(APPEND grpchdrs "${grpc_hdr}")
+    endforeach()
+  endmacro()
```

Comment:
We should turn this into a function that takes the library target to build and the list of protobuf files.
It would then be used like
```
add_library(whatever-proto STATIC)
addgrpcprotofiles(whatever-proto file1.proto file2.proto ...)
```
you'd just replace the `list()` operations by `target_sources(${target} PRIVATE ${proto_src} ${proto_hdr} ${grpc_src} ${grpc_hdr})`, change the parameter handling and make the macro into a function.